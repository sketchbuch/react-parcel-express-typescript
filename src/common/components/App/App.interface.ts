export interface Props {
  title: string;
}

export interface Store {
  app: boolean;
}
