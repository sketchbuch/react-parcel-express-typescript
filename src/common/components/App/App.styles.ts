import styled from 'styled-components';

export const StyledApp = styled.div`
  text-align: center;
  font-size: 2rem;
`;
StyledApp.displayName = 'StyledApp';

export const Para = styled.div`
  text-align: center;
  font-weight: bold;
  font-size: 1rem;
  padding-top: 20px;
`;
Para.displayName = 'Para';

export const LoadingMessage = styled(Para)`
  color: #999;
`;
LoadingMessage.displayName = 'LoadingMessage';

export const ErrorMessage = styled(Para)`
  color: red;
`;
ErrorMessage.displayName = 'ErrorMessage';
