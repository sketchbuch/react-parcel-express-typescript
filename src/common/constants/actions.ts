export const APP_LOADED = 'APP_LOADED';
