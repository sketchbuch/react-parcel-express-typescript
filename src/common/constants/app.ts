export const APP_TITLE = 'React Parcel Express Typescript'