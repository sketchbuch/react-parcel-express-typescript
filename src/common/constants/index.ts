export * from './actions';
export * from './routes';
export * from './server';
export * from './tests';
