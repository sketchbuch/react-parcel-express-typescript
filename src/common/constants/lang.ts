export const defaultLangKey = 'en';
export const availableLanguages = [defaultLangKey, 'de'] as const;
export const reactI18nextCookie = 'i18next';
export const lookupCookie = 'i18next';
export const lookupLocalStorage = 'i18nextLng';
export const lookupQuerystring = 'lng';