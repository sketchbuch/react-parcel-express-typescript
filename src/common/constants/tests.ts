export const BUNDLE_NAME = 'bundle_32463276.js';
