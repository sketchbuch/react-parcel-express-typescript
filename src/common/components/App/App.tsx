import React, { useEffect } from 'react';
import { Link, Switch, Route } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { Helmet } from 'react-helmet';
import { APP_LOADED } from '../../constants/actions';
import { ErrorMessage, LoadingMessage, Para, StyledApp } from './App.styles';
import { Props, Store } from './App.interface';
import { ROUTE_PAGE2, ROUTE_HOME } from '../../constants/routes';
import packageJson from '../../../../package.json';

export const Home: React.FC<{}> = () => (
  <React.Fragment>
    <Helmet>
      <title>App</title>
    </Helmet>
    <Para data-testid="app-home">
      <strong>
        <Link data-testid="app-home-link" to={ROUTE_PAGE2}>
          Go to page 2
        </Link>
      </strong>
    </Para>
  </React.Fragment>
);

export const NotFound: React.FC<{}> = () => (
  <React.Fragment>
    <Helmet>
      <title>NotFound - App</title>
    </Helmet>
    <ErrorMessage data-testid="app-404-errmsg">
      <strong>404</strong>
    </ErrorMessage>
    <Para data-testid="app-404-description">
      Sorry but the page you requested could not be found
    </Para>
    <Para data-testid="app-404">
      <strong>
        <Link data-testid="app-404-link" to={ROUTE_HOME}>
          Back to home
        </Link>
      </strong>
    </Para>
  </React.Fragment>
);

export const Page2: React.FC<{}> = () => (
  <React.Fragment>
    <Helmet>
      <title>Page 2 - App</title>
    </Helmet>
    <Para data-testid="app-page2">
      <strong>
        <Link data-testid="app-page2-link" to={ROUTE_HOME}>
          Back to home
        </Link>
      </strong>
    </Para>
  </React.Fragment>
);

const App: React.FC<Props> = ({ title }) => {
  const loaded: boolean = useSelector<Store, boolean>(state => state.app);
  const dispatch = useDispatch();

  useEffect(() => {
    const timer: number = window.setTimeout(() => dispatch({ type: APP_LOADED }), 1000);

    return (): void => clearTimeout(timer);
  }, []);

  return (
    <StyledApp>
      <Helmet>
        <title>App</title>
      </Helmet>
      <h1 data-testid="app-title">{title}</h1>
      <Para data-testid="app-description">
        v{packageJson.version} - Boilerplate for a webpack / koa based universal react app using
        babel, react-router, redux, and typescript
      </Para>

      {loaded === true ? (
        <LoadingMessage data-testid="app-loaded">Loaded!</LoadingMessage>
      ) : (
        <LoadingMessage data-testid="app-loading">Loading...</LoadingMessage>
      )}

      <Switch>
        <Route path={ROUTE_PAGE2} component={Page2} />
        <Route exact path={ROUTE_HOME} component={Home} />
        <Route component={NotFound} />
      </Switch>
    </StyledApp>
  );
};

export default App;
