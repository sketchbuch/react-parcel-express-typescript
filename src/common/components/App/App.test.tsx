import * as React from 'react';
import 'jest-styled-components';
import { wait } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { ErrorMessage, LoadingMessage, Para, StyledApp } from './App.styles';
import { Props } from './App.interface';
import { renderWithRedux, renderWithRouter } from '../../tests';
import { ROUTE_HOME, ROUTE_PAGE2 } from '../../constants';
import App, { Home, NotFound, Page2 } from './App';

describe('<App />', () => {
  const props: Props = {
    isSsr: false,
  };
  const LOADING_TXT = 'Loading...';
  const LOADED_TXT = 'Loaded!';

  /* test('Renders the title', () => {
    const { getByTestId, getByText } = renderWithRedux(<App {...props} />);
    expect(getByTestId('app-title')).toBeInTheDocument();
    expect(getByText(props.title)).toBeInTheDocument();
  }); */

  test('Renders loading message', () => {
    const { getByTestId, getByText, queryByTestId, queryByText } = renderWithRedux(
      <App {...props} />
    );

    expect(getByTestId('app-loading')).toBeInTheDocument();
    expect(queryByTestId('app-loaded')).toBeNull();
    expect(getByText(LOADING_TXT)).toBeInTheDocument();
    expect(queryByText(LOADED_TXT)).not.toBeInTheDocument();
  });

  test('Renders loaded message after 1 second', async () => {
    const { getByTestId, getByText, queryByTestId, queryByText } = renderWithRedux(
      <App {...props} />
    );
    await wait(
      () => {
        expect(queryByTestId('app-loading')).toBeNull();
        expect(getByTestId('app-loaded')).toBeInTheDocument();
        expect(queryByText(LOADING_TXT)).not.toBeInTheDocument();
        expect(getByText(LOADED_TXT)).toBeInTheDocument();
      },
      {
        timeout: 1000,
      }
    );
  });

  describe('Styled Components:', () => {
    test('<ErrorMessage />', () => {
      const tree = renderer.create(<ErrorMessage />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('<LoadingMessage />', () => {
      const tree = renderer.create(<LoadingMessage />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('<Para />', () => {
      const tree = renderer.create(<Para />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('<StyledApp />', () => {
      const tree = renderer.create(<StyledApp />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Routes:', () => {
    test('<Home /> renders correctly', () => {
      const { getByTestId } = renderWithRouter(<Home />);
      expect(getByTestId('app-home')).toBeInTheDocument();

      const linkChild = getByTestId('app-home-link');
      expect(linkChild).toBeInTheDocument();
      expect(linkChild).toHaveAttribute('href', ROUTE_PAGE2);
    });

    test('<NotFound /> renders correctly', () => {
      const { getByTestId } = renderWithRouter(<NotFound />);
      expect(getByTestId('app-404')).toBeInTheDocument();
      expect(getByTestId('app-404-description')).toBeInTheDocument();
      expect(getByTestId('app-404-errmsg')).toBeInTheDocument();

      const linkChild = getByTestId('app-404-link');
      expect(linkChild).toBeInTheDocument();
      expect(linkChild).toHaveAttribute('href', ROUTE_HOME);
    });

    test('<Page2 /> renders correctly', () => {
      const { getByTestId } = renderWithRouter(<Page2 />);
      expect(getByTestId('app-page2')).toBeInTheDocument();

      const linkChild = getByTestId('app-page2-link');
      expect(linkChild).toBeInTheDocument();
      expect(linkChild).toHaveAttribute('href', ROUTE_HOME);
    });
  });
});
