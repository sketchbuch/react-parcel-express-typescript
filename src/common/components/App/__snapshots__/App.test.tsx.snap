// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Styled Components: <ErrorMessage /> 1`] = `
.c0 {
  text-align: center;
  font-weight: bold;
  font-size: 1rem;
  padding-top: 20px;
  color: red;
}

<div
  className="sc-AxirZ c0"
/>
`;

exports[`<App /> Styled Components: <LoadingMessage /> 1`] = `
.c0 {
  text-align: center;
  font-weight: bold;
  font-size: 1rem;
  padding-top: 20px;
  color: #999;
}

<div
  className="sc-AxirZ c0"
/>
`;

exports[`<App /> Styled Components: <Para /> 1`] = `
.c0 {
  text-align: center;
  font-weight: bold;
  font-size: 1rem;
  padding-top: 20px;
}

<div
  className="c0"
/>
`;

exports[`<App /> Styled Components: <StyledApp /> 1`] = `
.c0 {
  text-align: center;
  font-size: 2rem;
}

<div
  className="c0"
/>
`;
